comp/autoscaling/datadogclient/impl/client.go

```diff
@@ -0,0 +1,124 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package datadogclientimpl implements the datadogclient component interface
+package datadogclientimpl
+
+import (
+	"sync"
+
+	datadogclient "github.com/DataDog/datadog-agent/comp/autoscaling/datadogclient/def"
+	configComponent "github.com/DataDog/datadog-agent/comp/core/config"
+	logComp "github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/comp/core/status"
+	compdef "github.com/DataDog/datadog-agent/comp/def"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"gopkg.in/zorkian/go-datadog-api.v2"
+)
+
+// Requires defines the dependencies for the datadogclient component
+type Requires struct {
+	Config configComponent.Component
+	Log    logComp.Component
+}
+
+// Provides defines the output of the datadogclient component
+type Provides struct {
+	compdef.Out
+	Comp           datadogclient.Component
+	StatusProvider status.InformationProvider
+}
+
+const (
+	metricsEndpointPrefix          = "https://api."
+	metricsEndpointConfig          = "external_metrics_provider.endpoint"
+	metricsRedundantEndpointConfig = "external_metrics_provider.endpoints"
+)
+
+// endpoint represent a datadog endpoint
+type endpoint struct {
+	Site   string `mapstructure:"site" json:"site" yaml:"site"`
+	URL    string `mapstructure:"url" json:"url" yaml:"url"`
+	APIKey string `mapstructure:"api_key" json:"api_key" yaml:"api_key"`
+	APPKey string `mapstructure:"app_key" json:"app_key" yaml:"app_key"`
+}
+
+// NewComponent creates a new datadogclient component
+func NewComponent(reqs Requires) (Provides, error) {
+	provides := Provides{}
+	client, err := createDatadogClient(reqs.Config)
+	if err != nil {
+		return provides, err
+	}
+	dc := &datadogClient{
+		datadogConfig:     reqs.Config,
+		client:            client,
+		numberOfRefreshes: 0,
+	}
+	// Register a callback to refresh the client when the api_key or app_key changes
+	reqs.Config.OnUpdate(func(setting string, _, _ any) {
+		if setting == "api_key" || setting == "app_key" {
+			dc.refreshClient()
+		}
+	})
+	provides.Comp = dc
+	provides.StatusProvider = status.NewInformationProvider(statusProvider{
+		dc: dc.client,
+	})
+	return provides, nil
+}
+
+// datadogClient is a wrapper around the datadog.Client, which allows for
+// refresh of the client pointer in case of app/api key changes
+type datadogClient struct {
```

Comment:
Having a struct and an import with the same name is confusing (ie import line 12). Could we rename one ?